src/libawkward/array/NumpyArray.cpp

```diff
@@ -624,16 +625,58 @@ namespace awkward {
     }
   }
 
-  const std::shared_ptr<Content> NumpyArray::flatten(int64_t axis) const {
-    if (axis != 0) {
-      throw std::runtime_error("FIXME: NumpyArray::flatten(axis != 0)");
+  const std::vector<ssize_t> flatten_shape(const std::vector<ssize_t>& shape, int64_t axis) {
+    if (axis >= shape.size()) {
+      throw std::invalid_argument(std::string("cannot flatten Numpy shape if axis > shape.size()"));
+    }
+    if (shape.size() == 1) {
+      return std::vector<ssize_t>();
+    }
+    else {
+      std::vector<ssize_t> out;
+      if (axis + 1 == shape.size()) {
+        out.emplace_back(std::accumulate(begin(shape), end(shape), 1, std::multiplies<ssize_t>()));
+      }
+      else {
+        const auto& indx = begin(shape) + axis;
+        if (indx > begin(shape)) {
+          out.insert(end(out), begin(shape), indx);
+        }
+        out.emplace_back(shape[axis]*shape[axis + 1]);
+        out.insert(end(out), indx + 2, end(shape));
+      }
+      return out;
+    }
+  }
+
+  const std::vector<ssize_t> flatten_strides(const std::vector<ssize_t>& strides, int64_t axis) {
+    if (axis >= strides.size()) {
+      throw std::invalid_argument(std::string("cannot flatten Numpy strides if axis > strides.size()"));
     }
+    if (strides.size() == 1) {
+      return std::vector<ssize_t>();
+    }
+    std::vector<ssize_t> out;
+    if (axis + 1 == strides.size()) {
+      out.emplace_back(strides[strides.size() - 1]);
+    }
+    else {
+      const auto indx = begin(strides) + axis;
+      if (indx > begin(strides)) {
+        out.insert(end(out), begin(strides), indx);
+      }
+      out.insert(end(out), indx + 1, end(strides));
+    }
+    return out;
+  }
```

Comment:
`flatten_shape` and `flatten_strides` used to be helper functions that were used deep in `getitem_*` methods. As long as they have the same behavior for the `axis=0` case so that `getitem_*` still works as before, they can be extended to be helper functions for the `flatten` method as well.

Oh—these `flatten_shape(std::vector<ssize_t>, int64_t)` and `flatten_stride(std::vector<ssize_t>, int64_t)` functions are defined _in addition_ to the old `flatten_shape(std::vector<ssize_t>)` and `flatten_stride(std::vector<ssize_t>)`. Is that why the many calls to them that do not specify an `axis` parameter aren't compilation errors? In that case, then, the old functions would be exactly the same as they used to be, and there's no reason to worry.

(Like, C++ isn't inserting a default `axis=0` when `flatten_shape` is called with one argument, right? I hope C++ is strict enough in its type check that it doesn't insert default arguments when there are too few of them, other than in `main`, that is.)

For the review, I should point out that your use of STL with `std::accumulate` and `emplace_back` exceeds my knowledge, but you're the expert here and the tests pass.  `:)`

Reply:
Yes, it's the function overloading feature of C++.